Mark and possibly block requests with AbstractRestProxyRequestHandler

matthiaso opened the pull request:
Not all resources in a backend application shall be directly accessible by browser; this commit breaks existing behavior, browser-accessible classes/methods must be marked with ApiExposed annotation otherwise they
 will be rejected with an HTTP 404 error if they are proxied by
 AbstractRestProxyRequestHandler.

419045

matthiaso committed:
Mark and possibly block proxied requests

Not all resources in a backend application shall be directly accessible
by browser; this commit breaks existing behavior, browser-accessible
classes/methods must be marked with ApiExposed annotation otherwise they
 will be rejected with an HTTP 404 error if they are proxied by
 AbstractRestProxyRequestHandler.

419045

matthiaso commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/ApiExposeHelper.java:
```diff
@@ -26,7 +27,7 @@ public class ApiExposeHelper {
-   * @return if the given instance has the {@link ApiExposed} annotation set (directly or on one of its super classes).
+   * @return if the given instance has the {@link ApiExposed} annotation with {@link ApiExposed#value()} = {@code true} set (directly or on one of its super classes).
    */
   public boolean hasApiExposedAnnotation(Object instance) {
```
rename to isApiExposed? also two other similar signatures.

paolobazzi commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/ApiExposedHelper.java:
```diff
@@ -26,28 +27,49 @@ public class ApiExposeHelper {
    */
-  public boolean hasApiExposedAnnotation(Class<?> clazz) {
-    return getAnnotation(clazz, ApiExposed.class) != null;
+  public boolean isApiExposed(Class<?> clazz) {
```
add javadoc for the following four methods

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/ApiExposedHelper.java:
```diff
@@ -26,28 +27,49 @@ public class ApiExposeHelper {
+    return ann != null && ann.value();
+  }
+
+  public boolean isApiExposed(Method method) {
```
include hint, that beside the method the annotation is also considered if put on the declaring class

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.api/src/main/java/org/eclipse/scout/rt/api/meta/MetaResource.java:
add `@ApiExposed` (useful for debugging)

paolobazzi commented on org.eclipse.scout.rt.api/src/main/java/org/eclipse/scout/rt/api/meta/MetaResource.java:
add `@ApiExposed` (useful for debugging)

paolobazzi commented on org.eclipse.scout.rt.api/src/main/java/org/eclipse/scout/rt/api/meta/MetaResource.java:
add `@ApiExposed` (useful for debugging)

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.api/src/main/java/org/eclipse/scout/rt/api/meta/MetaResource.java:
add `@ApiExposed` (useful for debugging)

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.api/src/main/java/org/eclipse/scout/rt/api/meta/MetaResource.java:
add `@ApiExposed` (useful for debugging)

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/cancellation/CancellationResource.java:
```diff
@@ -34,6 +35,7 @@ public class CancellationResource implements IRestResource {
 
   @PUT
   @Path("{requestId}")
+  @ApiExposed(false)
```
- do we add this annotation to each resource (even if false = default)?
- add javadoc indicating that this annotation shall not be removed?

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/servicetunnel/ProcessResource.java:
```diff
@@ -28,6 +29,7 @@ public class ProcessResource implements IRestResource {
   @POST
   @Consumes(MediaType.APPLICATION_OCTET_STREAM)
   @Produces(MediaType.APPLICATION_OCTET_STREAM)
+  @ApiExposed(false)
```
- do we add this annotation to each resource (even if false = default)?
- add javadoc indicating that this annotation shall not be removed?

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/doc/ApiDocGenerator.java:
```diff
@@ -238,7 +239,8 @@ protected boolean acceptRestResource(IRestResource res) {
     return Modifier.isPublic(m.getModifiers()) &&
-        !m.isAnnotationPresent(ApiDocIgnore.class);
+        !m.isAnnotationPresent(ApiDocIgnore.class) &&
+        BEANS.get(ApiExposeHelper.class).isApiExposed(m); // FIXME permission check?
```
add permission to overrule apiExposed flag -> e.g. ReadAllApiDocPermission?

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/a3cc68288dc0cb782cd10418693764f45cc3fd31..b38d059fe82f7363935dbdc83e8b4cf671275983

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/IApiExposedService.java:
```diff
@@ -0,0 +1,20 @@
+import org.eclipse.scout.rt.platform.ApplicationScoped;
+
+@ApplicationScoped
+public interface IApiExposedService {
```
add javadoc indicating purpose of this service

matthiaso replied:
remove service

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/ApiExposedFilter.java:
```diff
@@ -0,0 +1,78 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(ApiExposedFilter.class);
+
+  public static final String PROXIED_REQUEST_HEADER = "X-Scout-Proxied-Request";
```
we use `public static final String HTTP_HEADER_NAME` for similar header constants

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/ApiExposedFilter.java:
```diff
@@ -0,0 +1,78 @@
+
+@ApplicationScoped
+@Priority(Priorities.AUTHORIZATION)
+public class ApiExposedFilter implements ContainerRequestFilter {
```
Add javadoc similar to NoLogFilter: `Dynamically installed request filter that rejects proxied requests to methods which are not marked as API-exposed.

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/ApiExposedFilter.java:
```diff
@@ -0,0 +1,78 @@
+import org.slf4j.event.Level;
+
+@ApplicationScoped
+@Priority(Priorities.AUTHORIZATION)
```
switch the two annotations (e.g. order by length(annotation))

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/RestApplication.java:
```diff
@@ -106,6 +106,10 @@ public Map<String, Object> getProperties() {
     Set<Class<?>> contribute();
+
+    default Set<Class<?>> contribute(RestApplication application) {
+      return contribute();
```
where is this method used or invoked?

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/ExtRestApplication.java:
```diff
@@ -20,11 +20,11 @@ public class ExtRestApplication extends RestApplication {
   protected boolean isExtScope(Class<?> clazz) {
     RestApplicationScope annotation = clazz.getAnnotation(RestApplicationScope.class);
-    return annotation != null && ObjectUtility.isOneOf(RestApplicationScopes.EXT, annotation.value());
+    return (annotation == null && !IRestResource.class.isAssignableFrom(clazz)) || (annotation != null && ObjectUtility.isOneOf(RestApplicationScopes.EXT, annotation.value()));
```
this code snippet is a bit hard to read, suggestion:
```
  protected boolean isExtScope(Class<?> clazz) {
    RestApplicationScope annotation = clazz.getAnnotation(RestApplicationScope.class);
    if (annotation != null) {
      // check matching scope
      return ObjectUtility.isOneOf(RestApplicationScopes.EXT, annotation.value());
    }
    // include all non-IRestResource classes
    return !IRestResource.class.isAssignableFrom(clazz);
  }
```

matthiaso replied:
Revert as we do not register feature anymore for ext

matthiaso replied:
Reverted https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

matthiaso replied:
Do not revert, change is necessary otherwise `AbstractApiExposedFeature` is always installed also for this application (even though it should be ignored).

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/a3cc68288dc0cb782cd10418693764f45cc3fd31..b38d059fe82f7363935dbdc83e8b4cf671275983

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/AbstractApiExposedFeature.java:
```diff
@@ -0,0 +1,83 @@
+    }
+  }
+
+  protected abstract boolean isDefaultRestrict();
```
add javadoc

matthiaso replied:
Removed: https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/AbstractApiExposedFeature.java:
```diff
@@ -0,0 +1,83 @@
+    ApiExposed annotation = getApiExposedAnnotation(resourceInfo, context);
+
+    if (annotation == null && !isDefaultRestrict()) {
+      LOG.debug("ApiExposed annotation is not present, no {} installed for {}", ApiExposedFilter.class.getSimpleName(), resourceInfo);
```
Suggestion
[...] do not install {} [...]

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/AbstractApiExposedFeature.java:
```diff
@@ -0,0 +1,83 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractApiExposedFeature implements DynamicFeature {
```
Javadoc like "Abstract implementation of dynamic feature registering ApiExposedFilter for annotated methods/classes".

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/AbstractApiExposedFeature.java:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  @RestApplicationScope(RestApplicationScopes.EXT)
+  public static class ExtApiExposedFeature extends AbstractApiExposedFeature {
```
Could we also skip this feature completely for Ext scope by not registering ExtApiExposedFeature at all in ExtRestApplication (add to filterClass)?

matthiaso replied:
Remove, maybe also remove isDefaultRestrict

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/AbstractApiExposedFeature.java:
```diff
@@ -0,0 +1,83 @@
+      LOG.debug("ApiExposed annotation is not present, no {} installed for {}", ApiExposedFilter.class.getSimpleName(), resourceInfo);
+    }
+    else if ((annotation == null && isDefaultRestrict()) || !annotation.value()) {
+      // compute if operation is *not* api-exposed, if *not* exposed then install filter to restrict access
```
"compute" is already done when we reach this line of code. remove "compute" in this javadoc?

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/ApiExposeHelper.java:
Rename to ApiExpose**d**Helper?

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/AbstractApiExposedFeature.java:
```diff
@@ -0,0 +1,83 @@
+   * Get {@link ApiExposed} annotation; if provided on type and method, method wins.
+   */
+  protected ApiExposed getApiExposedAnnotation(ResourceInfo resourceInfo, FeatureContext context) {
+    return ObjectUtility.nvlOpt(resourceInfo.getResourceMethod().getAnnotation(ApiExposed.class), () -> BEANS.get(ApiExposeHelper.class).getApiExposedAnnotation(resourceInfo.getResourceClass()));
```
why not just (which also checks existence in declaring class?):
```
return BEANS.get(ApiExposeHelper.class).getApiExposedAnnotation(resourceInfo.getResourceMethod());
```

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/AdditionalHeadersHttpServletRequestWrapper.java:
```diff
@@ -0,0 +1,57 @@
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletRequestWrapper;
+
+public class AdditionalHeadersHttpServletRequestWrapper extends HttpServletRequestWrapper {
```
add javadoc, indicating purpose of class

matthiaso replied:
https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..a3cc68288dc0cb782cd10418693764f45cc3fd31

paolobazzi reviewed (commented):
Looks good, see inputs for some minor improvements.

matthiaso commented:
Changes since last review: https://github.com/eclipse-scout/scout.rt/compare/6afdcd507e7cd4fef7c7ed815fdbd73c632c1996..05bd36627367187ae926bd4ddf9297f6ec2176dc

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/doc/ReadAllApiDocPermission.java:
```diff
@@ -0,0 +1,22 @@
+import org.eclipse.scout.rt.api.data.security.PermissionId;
+import org.eclipse.scout.rt.security.AbstractPermission;
+
+public class ReadAllApiDocPermission extends AbstractPermission {
```
do we need a permission descriptor (NLS) here?

matthiaso replied:
I do not think so, there are other permissions in open source scout which are also not further described, e.g.: https://github.com/eclipse-scout/scout.rt/tree/c185cd793e622b400faa64ede0efcaa7322952e3/org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/security

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/AdditionalHeadersHttpServletRequestWrapper.java:
```diff
@@ -0,0 +1,60 @@
+import jakarta.servlet.http.HttpServletRequestWrapper;
+
+/**
+ * {@link HttpServletRequestWrapper}´to add additional headers for incoming requests.
```
```suggestion
 * {@link HttpServletRequestWrapper} to add additional headers for incoming requests.
```

paolobazzi commented on org.eclipse.scout.rt.ui.html.test/src/main/java/org/eclipse/scout/rt/ui/html/AbstractRestProxyRequestHandlerTest.java:
```diff
@@ -0,0 +1,51 @@
+import org.junit.Test;
+import org.mockito.internal.util.MockUtil;
+
+public abstract class AbstractRestProxyRequestHandlerTest {
```
javadoc? like Abstract test for {@link AbstractRestProxyRequestHandler} instances

paolobazzi commented on org.eclipse.scout.rt.ui.html.test/src/main/java/org/eclipse/scout/rt/ui/html/AbstractRestProxyRequestHandlerTest.java:
```diff
@@ -0,0 +1,51 @@
+import org.junit.Test;
+import org.mockito.internal.util.MockUtil;
+
+public abstract class AbstractRestProxyRequestHandlerTest {
```
will we have more impls. in future? or could this test be inlined in RestProxyRequestHandlerTest?

matthiaso replied:
Not sure, actually first I thought of writing a test for each impl but in the end I did just use a mock.

Pro for each impl: We ensure that header is added by every impl (even if methods are overwritten).

Contra for each impl: We need to have a test for each impl (not sure if it is really worth it), maybe even a new completness test (what about new impls added after my change).

matthiaso replied:
Future task: Inline it.

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/doc/ApiDocGenerator.java:
```diff
@@ -489,23 +506,28 @@ protected IHtmlDocument toHtml(String scope, List<ResourceDescriptor> resourceDe
+                        HTML.div(
+                            HTML.span("(" + CollectionUtility.format(r.getScopes()) + ")").cssClass("scope"),
+                            HTML.span(r.getPath()).cssClass("resource"),
+                            HTML.span(StringUtility.box("/", m.getPath(), "")).cssClass("method")).cssClass("path").toggleCssClass("not-api-exposed", !apiExposed))
```
Could we include this apiExposed Flag in MethodDescriptor to avoid recalculating the flag here?

matthiaso replied:
Future task: Include in MethodDescriptor if possible.

paolobazzi commented on org.eclipse.scout.rt.rest/src/main/java/org/eclipse/scout/rt/rest/resource/ApiExposedFilter.java:
```diff
@@ -0,0 +1,101 @@
+    boolean increaseLogLevel = !m_rejectRequests || Platform.get().inDevelopmentMode();
+    LOG.atLevel(increaseLogLevel ? Level.WARN : Level.DEBUG).log("External access to {} not allowed, reason: class/method is not marked with {}", getRequestPath(requestContext), ApiExposed.class.getSimpleName());
+    if (m_rejectRequests) {
+      requestContext.abortWith(BEANS.get(ErrorResponseBuilder.class).withHttpStatus(Response.Status.NOT_FOUND).withMessage(getNotFoundMessage(requestContext)).build());
```
to discuss: http code 404 vs. 401 (and no special handling for OPTIONS requests?)

matthiaso replied:
Future task: Possibly change to 401.

paolobazzi reviewed (commented):
Looks good, see comments for some minor inputs.
